### PR TITLE
Distribute entities during initial worldgen spawn

### PR DESCRIPTION
**Pull request by Foghrye4**
This is slightly controversial, because:
1. Algorithm does not match a vanilla algorithm serving a same purpose.
2. It does not resolve underlaying problem of a random numbers being not enough random.
Yet it doing a job of more or less natural looking animal spreading during worldgen. I also added a missing event (which is exist in vanilla code).
screenshot from 2018-09-30 18-12-44

### CODE REVIEW

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java [hunk 1]
```diff
@@ -57,14 +60,21 @@ public static void initialWorldGenSpawn(WorldServer world, Biome biome, int bloc
             Biome.SpawnListEntry currEntry = WeightedRandom.getRandomItem(world.rand, spawnList);
             int groupCount = MathHelper.getInt(random, currEntry.minGroupCount, currEntry.maxGroupCount);
             IEntityLivingData data = null;
-            int randX = blockX + random.nextInt(sizeX);
-            int randZ = blockZ + random.nextInt(sizeZ);
-
-            final int initRandX = randX;
-            final int initRandZ = randZ;
-
+            int randX = blockX + random.nextInt(sizeX) - 1;
```
Comment (Niko-sk2x) [inline src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java]: Why -1?

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java [hunk 2]
```diff
@@ -57,14 +60,21 @@ public static void initialWorldGenSpawn(WorldServer world, Biome biome, int bloc
...
                 for (int j = 0; j < 4; ++j) {
+                    randX += random.nextInt(5) + 1;
+                    randZ += random.nextInt(5) + 1;
+                    yaw += random.nextFloat() * 90.0f + 45.0f;
+                    if (randX >= blockX + sizeX)
+                        randX -= sizeX;
+                    if (randZ >= blockZ + sizeZ)
+                        randZ -= sizeZ;
+                    if (yaw > 360.0f)
+                        yaw -= 360.0f;
```
Comment (Niko-sk2x) [inline src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java]: Why the vanilla way of assigning random rotation wasn't enough?

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java [hunk 3]
```diff
@@ -57,14 +60,21 @@ public static void initialWorldGenSpawn(WorldServer world, Biome biome, int bloc
...
+            
             for (int i = 0; i < groupCount; ++i) {
                 for (int j = 0; j < 4; ++j) {
+                    randX += random.nextInt(5) + 1;
+                    randZ += random.nextInt(5) + 1;
+                    yaw += random.nextFloat() * 90.0f + 45.0f;
+                    if (randX >= blockX + sizeX)
+                        randX -= sizeX;
+                    if (randZ >= blockZ + sizeZ)
+                        randZ -= sizeZ;
```
Comment (Niko-sk2x) [inline src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java]: Why does that work and the vanilla way doesnt?

**Review by Niko-sk2x — COMMENTED** [review]
Overall I don't see much of a difference and how it could be noticeably different. Can you test it on a flat world and do side by side comparison?

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java [hunk 4]
```diff
@@ -81,18 +91,16 @@ public static void initialWorldGenSpawn(WorldServer world, Biome biome, int bloc
...
-                        spawnedEntity.setLocationAndAngles(randX + 0.5, pos.getY(), randZ + 0.5, random.nextFloat() * 360.0F, 0.0F);
+                        spawnedEntity.setLocationAndAngles(randX + 0.5, pos.getY(), randZ + 0.5, yaw, 0.0F);
+                        Event.Result forgeCanSpawn = ForgeEventFactory.canEntitySpawn(spawnedEntity, world, randX + 0.5f, (float) pos.getY(), randZ + 0.5f, null);
+                        if (forgeCanSpawn == Event.Result.DENY) {
+                            spawnedEntity.setDead();
+                            continue;
+                        }
                         world.spawnEntity(spawnedEntity);
                         data = spawnedEntity.onInitialSpawn(world.getDifficultyForLocation(new BlockPos(spawnedEntity)), data);
                         break;
```
Comment (Niko-sk2x) [inline src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java:100]: I now see why my code didn't work...

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java (hunk 1 above)
Comment (Foghrye4) [reply src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java]: I'm adding that 1 later so even if `random.nextInt(5)` return 0 position will be shifted anyway.

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java (hunk 2 above)
Comment (Foghrye4) [reply src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java]: I have no idea. By some reason all entities was looking in a same direction (-Z).

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java (hunk 3 above)
Comment (Foghrye4) [reply src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java]: An only option I can think out - random seed which was set based on coordinate hash has some weird properties.

File: src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java (hunk 4 above)
Comment (Foghrye4) [reply src/api/java/io/github/opencubicchunks/cubicchunks/api/worldgen/populator/WorldGenEntitySpawner.java:100]: Why?

**Comment by Foghrye4** [comment]
Rebased.

**Review by Niko-sk2x — APPROVED** [review]

**Pull request merged**